src/main/java/io/nats/client/impl/NatsMessage.java

```diff
@@ -13,57 +13,89 @@
 
 package io.nats.client.impl;
 
+import io.nats.client.Connection;
+import io.nats.client.Message;
+import io.nats.client.Subscription;
+
 import java.nio.ByteBuffer;
 import java.nio.CharBuffer;
 import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
 
-import io.nats.client.Connection;
-import io.nats.client.Message;
-import io.nats.client.Subscription;
-
 class NatsMessage implements Message {
-    private String sid;
+    public enum Kind {REGULAR, PROTOCOL, INCOMING}
+
+    private final Kind kind;
+
+    // Kind.REGULAR : just these fields
     private String subject;
     private String replyTo;
     private ByteBuffer data;
+    private boolean utf8mode;
+    private Headers headers;
+
+    // Kind.INCOMING : subject, replyTo, data and these fields
+    private String sid;

```

Comment:
I think it makes sense to get rid of strings as the internal data structures in `NatsMessage` and instead only convert to strings on demand when required by external consumers so that there's no string conversion overhead unless actually required by consumers. I've done this partially in #356. In addition I think we may want to move more parsing functionality from `NatsConnectionReader` into `NatsMessage` in general, in #356 I modified `NatsMessage` so that it is used to construct the complete outbound message line buffer required to write the message to the socket so that all message segment serialization functionality can be removed from `NatsConnectionWriter` I also have it so that the primary `NatsMessage` message buffer is also used as the backing buffer for the more specific buffers for the various message segments, this allows for directly calling equals operations on the segment buffers. Something along these lines may make sense for handling inbound messages we well, essentially where we use a single primary buffer with other referencing buffers, although they could also be kept separate for inbound as inbound messages do not ever need to be serialized back to full message lines like outbound messages need to be.